### ISSUE TITLE
Is there an issue with animation on Ipad Pro after swipe/drag/pinch?

### ISSUE THREAD

**Issue by nikreations** [openseadragon/openseadragon#1753]
I hope someone can shed some light on this..

I have been battling with this issue for the past many days, trying out different options without any progress. Basically the animations (zoom, panning etc) work perfectly smooth on android devices and windows. But on Ipad pro (11, with Ipad IOS 13.2.3 Chrome & Safari), as soon as I do a swipe or drag or pinch, all animations stop being smooth. If I don't use the swipe/drag/pinch, but just the buttons, the animation is butter smooth. I have setup a test here: https://in360degrees.co.uk/test/test-ipad-pan.html 

The image in the test link is 91424 x 11280 My working image (which I can't display publicly at the moment) is going to be 27984 x 60000 and a swipe action on this image is extremely jerky (like it has performance issues) on IPad pro (Safari and Chrome). My android phone (Samsung s8) has no issues whatsoever.

Things I have tried without success:
-Setting canvas-drag preventDefaultAction to true for IPad only, thinking it was the native Ipad gestures causing issues. Even when I do this, the action of swipe, although no animation takes place, makes the other animations go jerky.
-Trawled through issues here regarding ipad and ios, but nothing similar to what I am facing.
-Changed from zoomifytileservice to dzi

**Comment by nikreations** [comment]
Some more info... I set the width and height of the canvas to 50% and the issue disappears. I tried with 98% but the issue still remains. So I incrementally reduced the width and height, until 50% gave me a good smooth animation for all actions. I still need the canvas to be 100%. So still in the dark as to what is causing this issue on ipad pro. I haven't got any other ios devices to test so not sure if this issue is specific to ios devices or only ipad pros.

**Comment by nikreations** [comment]
I have setup 2 links for testing (has a blurred image but its the original size 27984 x 60000):
canvas size is 50%: http://in360degrees.co.uk/indie360/testOSD/ipad-test-pan-50.html
canvas size is 100%: http://in360degrees.co.uk/indie360/testOSD/ipad-test-pan-100.html

**Comment by iangilman** [comment]
Strange that it's just the gestures and not the buttons. I wonder if there's a clue there.

I've tested your pages (thank you for sharing them!) on an iPad Air running iOS 12.4.2, and all three seem smooth (I mean, it's not 60fps, but it's acceptable, and there doesn't seem to be any difference between the gestures and the buttons). The Brighton Beach image seems more responsive than the other two, FWIW; presumably that's just an OSD viewer without any other stuff? I also tried on an iPhone 8 running iOS 13.1.3, and they all seemed smooth there as well. This makes me think it's probably the iPad Pro specifically, perhaps the screen size, that's causing the issue. Unfortunately I don't have access to an iPad Pro to test with.

Have you tried other OSD sites on that device? Do they all have the same issue? For instance, this is a big image and it's full screen: 

http://www.in2white.com/

Anyway, I'm not sure what to suggest. If it's a limitation in the iPad Pro, I guess we need to learn more about what's causing it.

**Comment by nikreations** [comment]
Thanks for your input Ian. I tested in2white. It is a similar experience. It works smoothly first, then the jerky panning and zooming starts. I am going to source another device and check if it is not just my device. I don't think it is a performance issue because it works nicely when using just the buttons, its something in the swipe gesture in the 100% version that causes the jerky behaviour (50% version works perfect). I will record a video by end of today.

**Comment by iangilman** [comment]
Hmm... I wonder if it's maxing out the memory of the browser. It might be worth trying setting `maxImageCacheCount` to something smaller than the default 200. 

We seem to have a number of iPad related issues being reported lately. See https://github.com/openseadragon/openseadragon/issues/1662, https://github.com/openseadragon/openseadragon/issues/1724, https://github.com/openseadragon/openseadragon/issues/1728, and https://github.com/openseadragon/openseadragon/issues/1751.

There's also the patch https://github.com/openseadragon/openseadragon/pull/1754 that has just landed on master.

**Comment by nikreations** [comment]
Thanks Ian, I tried setting maxImageCacheCount to 50, 100 and 500. I also tried the patch, but it is exactly the same jerky experience for pinch (zoom) and swipe(pan) (made no difference when setting to different values). I did get to test on other Ipads briefly. Ipad 11 (another device to mine) & Ipad 10.2 (2019) shows the same jerky experience. However, this is the interesting bit... Ipad 12.9, Ipad Air & Ipad mini, the pinch and swipe actions are perfectly smooth. Since other things work good, I am thinking of leaving this issue for now.

**Comment by iangilman** [comment]
Fair enough. I'm out of ideas at the moment. I'll let you know if something comes up!

**Comment by nickngsr** [comment]
I'm getting this issue in safari on a mac also as well as a few iOS devices, on initial load zooming is perfectly smooth, after panning it starts to stutter and zooming also now stutters. Cpu usage suffers tremendously at this point

Typical CPU usage: 
Before panning, zooming utilises 19% on 1 core
After panning, zooming utilises 100% cpu on 1 core

I have tested it on other sites with OSD on, including the in2white linked above, no zoom but the effect is visible while panning

**Comment by nickngsr** [comment]
The xkcd example referenced in #1408 using the very old version does not have this problem

**Comment by nickngsr** [comment]
Resizing the canvas seems to reset it until you pan again

**Comment by nickngsr** [comment]
It appears to be that safari can't handle the canvas having focus. The performance issues happen as soon as https://github.com/openseadragon/openseadragon/blob/0383f8fbeb7b36921e5faebe8e3eaf8de1f7bae7/src/viewer.js#L2679
is called. Preventing the canvas from having focus stops the issue (but adds a new one for keyboard shortcuts)

**Comment by iangilman** [comment]
Interesting! Might it be possible to attach the keyboard event handler to another element in the viewer and not get this effect?

At the very least, we could not do the focus when on a mobile device, assuming there is no need for keyboard input there.

Note, by the way, that the viewer's `canvas` property is just a div, not a true HTML canvas (for confusing historical reasons).

**Comment by msalsbery** [comment]
@nikreations 

I'm finishing up an event-handling (MouseTracker) patch and I'm curious if the new code improves the issues you're seeing.

If you have a little time to try it out, I would love to hear your feedback!

[Links to Demos and Preview Builds](https://github.com/openseadragon/openseadragon/pull/1872#issuecomment-673608460)

Cheers

**Comment by msalsbery** [comment]
> with version 2.5 the hotspots became unclickable/untapable on all devices

@nikreations  I just wanted to check in so you'd know we haven't forgotten about this issue. Been busy working on v2.5.0, but when it lands I can work with you to try to figure out what's going on.

I've profiled the site you linked to (_**very nice looking site by the way!**_), and the DOM looks pretty simple so there must be a lot of javascript going on. The site gets sluggish quickly on mobile devices, but also doesn't take long to slow even on a desktop browser. While it doesn't appear too busy when looking at the page, profiling shows slowing framerates and steadily increasing memory usage.

I didn't spend too much time profiling memory, but it also appears to leak a lot of objects. Of course I don't know the code so I can't be certain if it's ok or not.

These are just my initial observations and don't mean much as I don't know the code. But clearly there's room for improvement with performance (why you filed this issue ;-)) and I do want to get to the bottom of this to make sure OpenSeadragon is working correctly and also the documentation is sufficient so we as end-users can use the library efficiently.

**Comment by iangilman** [comment]
> Additionally, with version 2.5 the hotspots became unclickable/untapable on all devices.

Probably worth looking into this aspect before merging #1872, to make sure we are not introducing new bugs there.

@nikreations I agree it's a great site! Do you mind if I share it?

**Comment by msalsbery** [comment]
> Probably worth looking into this aspect before merging #1872

Yep....really need some feedback here - I can't debug without seeing how it's being used!

cc @nikreations

**Comment by msalsbery** [comment]
@nikreations 

How do these sites behave on Ipad Pro 11?

* [OpenSeadragon Imaging ImagingHelper Demo](https://openseadragon-imaging.github.io/#/imaginghelper)
* [OpenSeadragon Imaging ConsoleHook Demo](https://openseadragon-imaging.github.io/#/consolehook)

If you have a few minutes, could you please (on an Ipad Pro 11) [go to this site](https://openseadragon-imaging.github.io/#/imaginghelper) and copy the navigator, osdBrowser, and osdMouseTracker text from the properties pane? It should look something like this:
```
navigator: {
  "appVersion": "5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36 Edg/85.0.564.44",
  "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36 Edg/85.0.564.44"
}
osdBrowser: {
  "vendor": "Chromium Edge",
  "version": 85,
  "alpha": true,
  "opacity": true
}
osdMouseTracker: {
  "wheelEventName": "wheel",
  "havePointerCapture": true,
  "havePointerEvents": true,
  "unprefixedPointerEvents": true,
  "havePointerOverOut": true
}
```
Thanks!

**Comment by nikreations** [comment]
HiHere is the info

   
   -    navigator: {
  "appVersion": "5.0 (iPad; CPU OS 13_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) CriOS/84.0.4147.122 Mobile/15E148 Safari/604.1",
  "userAgent": "Mozilla/5.0 (iPad; CPU OS 13_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) CriOS/84.0.4147.122 Mobile/15E148 Safari/604.1"
}
   -    osdBrowser: {
  "vendor": "Safari",
  "version": 1.5e+149,
  "alpha": true,
  "opacity": true
}
   -    osdMouseTracker: {
  "wheelEventName": "wheel",
  "havePointerCapture": true,
  "havePointerEvents": true,
  "unprefixedPointerEvents": true,
  "havePointerOverOut": true
}

    On Thursday, 3 September 2020, 00:41:48 BST, Mark Salsbery <notifications@github.com> wrote:  
 
 


@nikreations

How do these sites behave on Ipad Pro 11?
   
   - OpenSeadragon Imaging ImagingHelper Demo
   - OpenSeadragon Imaging ConsoleHook Demo

If you have a few minutes, could you please (on an Ipad Pro 11) go to this site and copy the navigator, osdBrowser, and osdMouseTracker text from the properties pane? It should look something like this:
navigator: {
  "appVersion": "5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36 Edg/85.0.564.44",
  "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36 Edg/85.0.564.44"
}
osdBrowser: {
  "vendor": "Chromium Edge",
  "version": 85,
  "alpha": true,
  "opacity": true
}
osdMouseTracker: {
  "wheelEventName": "wheel",
  "havePointerCapture": true,
  "havePointerEvents": true,
  "unprefixedPointerEvents": true,
  "havePointerOverOut": true
}

Thanks!

**Comment by nikreations** [comment]
I tried the 2 links. Not smooth on ipad pro 11. Buttery smooth on desktop (windows 10, chrome) and on my Samsung S8 (Chrome).

**Comment by nikreations** [comment]
> > Additionally, with version 2.5 the hotspots became unclickable/untapable on all devices.
> 
> Probably worth looking into this aspect before merging #1872, to make sure we are not introducing new bugs there.
> 
> @nikreations I agree it's a great site! Do you mind if I share it?

Thanks @iangilman The direct link to OSD part I posted is part of a bigger project https://virtualtour.ornc.org/ which can be shared, but unfortunately not the direct link to OSD part (yet). There are 2 versions, English & British Sign language.

**Comment by msalsbery** [comment]
I> I tried the 2 links. Not smooth on ipad pro 11. Buttery smooth on desktop (windows 10, chrome) and on my Samsung S8 (Chrome).
> 
> THanks for looking into this issue.

@nikreations thank you! 

So 2.4.1 smooth, 2.4.2 not smooth, and 2.5.0 (those linked pages) not smooth, correct?

**Comment by msalsbery** [comment]
@nikreations 

What about demos here https://openseadragon.github.io/

Thank you for the feedback! I really wish I had the device in my hands

**Comment by msalsbery** [comment]
@nikreations do you have an iPhone? If so what are you seeing there? I ask because on Webkit browsers is where I see the worst “smoothness” and I’m trying to get a feel for what you’re seeing.

At those two links to my demos were you able to pan and switch images with the buttons?

Leaving this here for my notes...
```
iPhone 7 13.6.1

navigator: {
  "appVersion": "5.0 (iPhone; CPU iPhone OS 13_6_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.1.2 Mobile/15E148 Safari/604.1",
  "userAgent": "Mozilla/5.0 (iPhone; CPU iPhone OS 13_6_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.1.2 Mobile/15E148 Safari/604.1"
}
osdBrowser: {
  "vendor": "Safari",
  "version": 1.5e+149,
  "alpha": true,
  "opacity": true
}
osdMouseTracker: {
  "wheelEventName": "wheel",
  "havePointerCapture": true,
  "havePointerEvents": true,
  "unprefixedPointerEvents": true,
  "havePointerOverOut": true
}
```

**Comment by nikreations** [comment]
No sorry, I don't have an iphone. The demos at https://openseadragon.github.io/ also show the same behaviour. I haven't checked it now, but before I got to check on other Ipads (pro 11, pro 12.9, Air, Mini & 10.2). 12.9, Air & Mini had no issue. It was just Ipad 10.2 and Pro 11, mine and another device. I checked my project and in2white

The webpage's "view source" should display all the code there is running on that page. (view-source:https://virtualtour.ornc.org/virtualtour-std/ceiling/). I am using XML to store all the hotspot details, there is jquery for hiding and showing the info panels on top of OSD. The info panels are iframes

Basically I am not a mainstream programmer, my main job is photography. So something like this is way beyond my understanding, but if there is anyway I can provide debugging info, just let me know.

**Comment by msalsbery** [comment]
@nikreations  Thanks again...I'll definitely be asking more from you! ;-)

I've moved over to a Mac, where fortunately (or unfortunately, depending on how I look at it) OpenSeadragon isn't so smooth on desktop Webkit (Safari) either. I'll be taking a look at 2.4.1, 2.4.2, and 2.5.0 and hopefully find out what's up

**Comment by iangilman** [comment]
> Thanks @iangilman The direct link to OSD part I posted is part of a bigger project https://virtualtour.ornc.org/ which can be shared, but unfortunately not the direct link to OSD part (yet). There are 2 versions, English & British Sign language.

Cool. Well, I'd love to share it when it's ready! Please let me know here, or [on Twitter](https://twitter.com/openseadragon), or, if you're up for it, make a pull request on this file:

https://github.com/openseadragon/site-build/blob/master/www/in-the-wild.html

**Comment by msalsbery** [comment]
@nikreations 

Now that #1872 has landed it may be worth retesting with the latest OpenSeadragon build...

**Comment by nikreations** [comment]
Thanks @msalsbery I just quickly replaced the js with the build and tested. Clicking/tapping on overlays did not work (same as 2.5). So I can't check if the initial problem still exists. But I have just tested 2.4.2 again and it seems the issue I reported is almost gone!! Only if I was nitpicking, I would say just some slight judder when swiped or zoomed but very much acceptable. It has been a while since this project was implemented and my IPad pro 11 is now on IOS14.4 Could it be that problem was due to apple's code rather than OSD's since there were no such issues on other Ipad models (except 11 & 10.2), android and windows? Happy to provide further info.

**Comment by msalsbery** [comment]
@nikreations  Thanks for the quick reply! Where did you get the build? I don't think it's officially up yet as we had an issue with tests not passing yesterday (I'm waiting for @iangilman to get the fixed tests merged in and do a build ;-))

**Comment by nikreations** [comment]
Oh I see, It was one of the links in #1872 https://openseadragon-imaging.github.io/old-demo/lib/openseadragon.min.js
It had a build date of 2nd March 2021.

**Comment by msalsbery** [comment]
@nikreations  Ok cool...that should be pretty close to the final version (I haven't updated the builds since the PR landed, yet)

**Comment by msalsbery** [comment]
@nikreations 

Regarding clicking/tapping on overlays...[here's a CodePen](https://codepen.io/msalsbery/pen/dyOeXqO) that shows a new way of using MouseTracker on overlays and controlling the bubbling/disposition of events

**Comment by amroc-dev** [comment]
Just wanted to say I've noticed the slow down problem on my iPhone Safari. I've noticed that the slowdown always begins after you have fully released all touches *for the first time*. So with your very first interaction you can pinch with two fingers, then leave one finger on the screen, pan around, zoom again with two fingers, and so on, and so long as you leave at least one finger touching, the frame rate stays at full 60fps. However as soon as you fully release, from then on everything will update at 30fps.

**Comment by iangilman** [comment]
@amroc-dev thank you for the report! Is that with the current release (e.g. what's on http://openseadragon.github.io/) or with the latest master since PR #1872 was merged?

**Comment by msalsbery** [comment]
This may be related to a Webkit bug...I'm documenting it and working on a fix, will link this issue

**Comment by amroc-dev** [comment]
> @amroc-dev thank you for the report! Is that with the current release (e.g. what's on http://openseadragon.github.io/) or with the latest master since PR #1872 was merged?

Current release yes. 2.4.2, installed via npm. Thx!

**Comment by amroc-dev** [comment]
> This may be related to a Webkit bug, see #1962
> 
> Working on a workaround...

Brilliant, thank you.

**Comment by msalsbery** [comment]
> > @amroc-dev thank you for the report! Is that with the current release (e.g. what's on http://openseadragon.github.io/) or with the latest master since PR #1872 was merged?
> 
> Current release yes. 2.4.2, installed via npm. Thx!

@amroc-dev 

The WebKit bug affects overlaid elements on the viewer but should be ok with a standalone viewer. 

Have you tried the preview builds of 3.0.0 yet? Curious if there’s improvements there...[you can get builds and links to demos here](https://github.com/openseadragon/openseadragon/pull/1872#issuecomment-673608460). 

I do want to get a workaround in for the WebPack issue before we release 3.0.0...

Thank you for the feedback!

**Comment by amroc-dev** [comment]
@msalsbery I haven't, I will do and report back. Thanks again :)

**Comment by amroc-dev** [comment]
@msalsbery Ok I haven't yet tried the new build with my project, because I tried one of the demo links (https://openseadragon-imaging.github.io/#/imaginghelper), and I can confirm that exhibits the same slowdown behaviour under the same conditions I described earlier.

**Comment by SamHall** [comment]
Checking in to see if anyone has found a workaround for the iPad 'bad behavior'. 
It's been a thorn in my side for a long time. 
My workaround has been to present Openseadragon in modals and disabling navigation buttons on all but larger screens.
Example:  [https://chattanoogahistory.com/epb_archive2](url)
I'm using Bootstrap 4 as the framework.
However, I'd prefer presenting directly. Perhaps some smart person here can look at the modal presentation approach and determine what secret sauce could be extracted to make full-screen behave on iPad?

**Comment by SamHall** [comment]
I found that this page also seems to work well on iPad/full screen but is NOT in a modal. 
Perhaps there’s a clue within how I’ve presented it?  

https://chattanoogahistory.com/williamspapers

**Comment by amroc-dev** [comment]
> I found that this page also seems to work well on iPad/full screen but is NOT in a modal...

I would love an answer for this too. If there is just a presentation method that can sidestep the issue.

**Comment by amroc-dev** [comment]
Think I've found a workaround based on a comment earlier in this thread. In your OpenSeaDragon element, respond to both onClick (for desktop Safari), and onTouchStart (for iOS safari) events, and then when they trigger call blur() on OpenSeaDragon's canvas. Seems to fix the slow down.

**Comment by msalsbery** [comment]
@amroc-dev  Thank you for that...that may be an important clue!

This issue is NOT related to the Webkit bug I mentioned above...I can clearly see the performance drop after a single click in Safari desktop on a Mac

I'm looking into this now so I'll try your workaround as well, thanks!

**Comment by msalsbery** [comment]
@amroc-dev 

I confirmed that it is when the OSD canvas div gets focus the performance drops. Great find, thank you.

Your workaround is cool, with the only issue being no keyboard navigation in the viewer.

I'm still looking into why it happens on Webkit...seems pretty ridiculous, but maybe there's something there that can be turned off with CSS or something. Fingers crossed...

**Comment by msalsbery** [comment]
@amroc-dev  @SamHall 

Try this in css:
```javascript
.openseadragon-canvas:focus {
  outline-style: none;
}
```
Webkit desktop (Safari on Mac) sets :focus for the div to
```javascript
{
  outline-color: -webkit-focus-ring-color;
  outline-style: auto;
  outline-width: 5px;
}
```
If it's not my imagination, the lag seems to go away by removing the focused outline-style

**Comment by SamHall** [comment]
UPDATE:  The behavior I'm seeing appears to only be happening in Safari on iPad, Chrome is not having the issue of 'popping out' of fullscreen mode. I realize this is not necessarily related to the Webkit bug. 
And the CSS does not appear to affect this unfortunately. 
Safari places an 'X' close icon in the upper left corner while in fullscreen content. I wonder if that is being triggered while interacting with the canvas?

**Comment by msalsbery** [comment]
@SamHall 

Do you have a separate issue created for the behavior you're seeing? If not, could you please open one? :)

I don't think your issue is related to this issue, which is the poor pan/zoom animation performance on WebKit browsers after an initial drag/click gesture occurs on the viewer...

**Comment by SamHall** [comment]
> @SamHall
> 
> Do you have a separate issue created for the behavior you're seeing? If not, could you please open one? :)
> 
> I don't think your issue is related to this issue, which is the poor pan/zoom animation performance on WebKit browsers after an initial drag/click gesture occurs on the viewer...

I will open a new issue. My apologies - not meaning to hijack this one. When I posted earlier I misinterpreted this issue for the one I was having.

**Comment by msalsbery** [comment]
> I will open a new issue. My apologies - not meaning to hijack this one. When I posted earlier I misinterpreted this issue for the one I was having.

No problem at all...just want to make sure your issue gets seen by the right people who know the full-screen/full-page code!
